internal/statestorage/redis/redispb/matchobject.go

```diff
@@ -0,0 +1,143 @@
+)
+
+// Logrus structured logging setup
+var (
+	moLogFields = log.Fields{
+		"app":       "openmatch",
+		"component": "statestorage",
+	}
+	moLog = log.WithFields(moLogFields)
+)
+
+// UnmarshalFromRedis unmarshals a MatchObject from a redis hash.
+// This can probably be made generic to work with other pb messages in the future.
+// In every case where we don't get an update, we return an error.
+func UnmarshalFromRedis(ctx context.Context, pool *redis.Pool, pb *om_messages.MatchObject) error {
+
+	// Get the Redis connection.
+	redisConn, err := pool.GetContext(context.Background())
+	defer redisConn.Close()
+	if err != nil {
+		moLog.WithFields(log.Fields{
+			"error":     err.Error(),
+			"component": "statestorage",
+		}).Error("failed to connect to redis")
+		return err
+	}
+
+	// Prepare redis command.
+	cmd := "HGETALL"
+	key := pb.Id
+	resultLog := moLog.WithFields(log.Fields{
+		"component": "statestorage",
+		"cmd":       cmd,
+		"key":       key,
+	})
+	pbMap, err := redis.StringMap(redisConn.Do(cmd, key))
+	if len(pbMap) == 0 {
+		return errors.New("matchobject key does not exist")
+	}
+
+	// Put values from redis into the MatchObject message
+	pb.Error = pbMap["error"]
+	pb.Properties = pbMap["properties"]
+
+	// TODO: Room for improvement here.
+	if j := pbMap["pools"]; j != "" {
+		poolsJSON := fmt.Sprintf("{\"pools\": %v}", j)
+		err = jsonpb.UnmarshalString(poolsJSON, pb)
+		if err != nil {
+			resultLog.Error("failure on pool")
+			resultLog.Error(j)
+			resultLog.Error(err)
+		}
+	}
+
+	if j := pbMap["rosters"]; j != "" {
+		rostersJSON := fmt.Sprintf("{\"rosters\": %v}", j)
+		err = jsonpb.UnmarshalString(rostersJSON, pb)
+		if err != nil {
+			resultLog.Error("failure on roster")
+			resultLog.Error(j)
+			resultLog.Error(err)
+		}
+	}
+	moLog.Debug("Final pb:")
+	moLog.Debug(pb)
+	return err
+}
+
+// Watcher makes a channel and returns it immediately.  It also launches an
+// asynchronous goroutine that watches a redis key and returns updates to
+// that key on the channel.
+//
+// The pattern for this function is from 'Go Concurrency Patterns', it is a function
+// that wraps a closure goroutine, and returns a channel.
+// reference: https://talks.golang.org/2012/concurrency.slide#25
+func Watcher(ctx context.Context, pool *redis.Pool, pb om_messages.MatchObject) <-chan om_messages.MatchObject {
+
+	watchChan := make(chan om_messages.MatchObject)
+	results := om_messages.MatchObject{Id: pb.Id}
+
+	go func() {
+		// var declaration
+		var err = errors.New("haven't queried Redis yet")
+
+		// Loop, querying redis until this key has a value
+		for err != nil {
+			select {
+			case <-ctx.Done():
+				// Cleanup
+				close(watchChan)
+				return
+			default:
+				//results, err = Retrieve(ctx, pool, key)
+				results = om_messages.MatchObject{Id: pb.Id}
+				err = UnmarshalFromRedis(ctx, pool, &results)
+				if err != nil {
+					moLog.Debug("No new results")
+					time.Sleep(2 * time.Second) // TODO: exp bo + jitter
```

Comment:
If we make this a `rate` parameter or a `rateChan` parameter, then the caller can describe the jitter and frequency for the queries.

Reply:
[Issue 59](https://github.com/GoogleCloudPlatform/open-match/issues/59)